langchain_tools_demo/agent.py

```diff
@@ -49,12 +48,12 @@ def __init__(self, client, agent) -> None:
 user_agents: Dict[str, UserAgent] = {}
 
 
-def get_id_token(url: str) -> str:
+def get_id_token() -> str:
     """Helper method to generate ID tokens for authenticated requests"""
     # Use Application Default Credentials on Cloud Run
     if os.getenv("K_SERVICE"):
         auth_req = google.auth.transport.requests.Request()
```

Comment:
```suggestion
id_token = None # (token, expiry_time)

def get_id_token() -> str:
    """Helper method to generate ID tokens for authenticated requests"""
       if id_token is None or id_token[1] <= time.now():
           auth_req = google.auth.transport.requests.Request()
           token =  google.oauth2.id_token.fetch_id_token(auth_req, BASE_URL)
           id_token = (token, datetime.now() + timedelta(minutes=55))
```

Reply:
We still need to use the shell command 'gcloud auth print-identity-token` because `gcloud auth login` does not set Application Default Credential: https://www.googlecloudcommunity.com/gc/Serverless/fetch-id-token-or-equivalent-locally-with-gcloud-non-default/m-p/552409